server/src/main/java/keywhiz/service/daos/SecretContentDAO.java

```diff
@@ -75,6 +75,16 @@ public long createSecretContent(long secretId, String encryptedContent, String h
       String creator, Map<String, String> metadata, long expiry, long now) {
     SecretsContentRecord r = dslContext.newRecord(SECRETS_CONTENT);
 
+    if (config.getMaximumSecretSizeInBytes() != null) {
+      if (encryptedContent.length() > config.getMaximumSecretSizeInBytes()) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Secret is too large. Secret size %d bytes exceeds maximum size of %d bytes",
+                encryptedContent.length(),
+                config.getMaximumSecretSizeInBytes()));
+      }
+    }
```

Comment:
This is the actual behavior change. In hindsight, it probably makes sense to push it down into its own method so that this one doesn't get too cluttered.

In keeping with existing validation for secret series names, this logic is specified at the DAO level since it's a chokepoint for various APIs.

Reply:
I'll trace through the call stack to see how the secret comes in. IIRC we're already expecting it to be encoded in some fashion but I want to be sure.

Reply:
Our server-side APIs declare that the secret content is @ValidBase64, and in the Keywhiz CLI client there's a line which is responsible for transforming the raw secret into base64 before sending it across the wire. I'll put together an error message which takes this into account and also provides a more useful hint as to the acceptable raw secret size.

Reply:
Upon further reflection, I think moving the validation before encryption happens will allow us to more accurately reflect the maximum secret size back to the caller. Fortunately there is another chokepoint, SecretController, that I'd missed previously. Moving the logic there makes this a better user experience and also simplifies testing somewhat.

Stay tuned.